### ISSUE TITLE
Bridge Crashes When Counter Decreases

### ISSUE THREAD

**Issue by ekristen** [prometheus/statsd_exporter#11]
Basically I'm forwarding `nsq` stats to statsd, with statsd repeating to the bridge. When NSQ restarts its stats restart, or when a NSQ reader disconnects and reconnects the stats reset, therefore the counter would reset.

I was told in IRC that Prometheus itself understands concepts like this. 

Is there a way we can make the bridge not crash in this scenario?

```
2015/04/10 23:17:12 Starting StatsD -> Prometheus Bridge...
2015/04/10 23:17:12 Accepting StatsD Traffic on :9125
2015/04/10 23:17:12 Accepting Prometheus Requests on :9102
panic: counter cannot decrease in value

goroutine 16 [running]:
runtime.panic(0x6fef00, 0xc20870fb10)
        /usr/src/go/src/pkg/runtime/panic.c:279 +0xf5
github.com/prometheus/client_golang/prometheus.(*counter).Add(0xc208196200, 0xc0ef876000000000)
        /go/src/github.com/prometheus/client_golang/prometheus/counter.go:69 +0xf9
main.(*Bridge).Listen(0xc20800f500, 0xc2080a2000)
        /go/src/app/bridge.go:217 +0x617
main.main()
        /go/src/app/main.go:127 +0x5e3

goroutine 19 [finalizer wait, 8 minutes]:
runtime.park(0x41a430, 0x952a78, 0x950cc9)
        /usr/src/go/src/pkg/runtime/proc.c:1369 +0x89
runtime.parkunlock(0x952a78, 0x950cc9)
        /usr/src/go/src/pkg/runtime/proc.c:1385 +0x3b
runfinq()
        /usr/src/go/src/pkg/runtime/mgc0.c:2644 +0xcf
runtime.goexit()
        /usr/src/go/src/pkg/runtime/proc.c:1445

goroutine 26 [IO wait, 5 minutes]:
net.runtime_pollWait(0x7f1489fffb20, 0x72, 0x0)
        /usr/src/go/src/pkg/runtime/netpoll.goc:146 +0x66
net.(*pollDesc).Wait(0xc2080ba140, 0x72, 0x0, 0x0)
        /usr/src/go/src/pkg/net/fd_poll_runtime.go:84 +0x46
net.(*pollDesc).WaitRead(0xc2080ba140, 0x0, 0x0)
        /usr/src/go/src/pkg/net/fd_poll_runtime.go:89 +0x42
net.(*netFD).accept(0xc2080ba0e0, 0x823cd8, 0x0, 0x7f1489ffe520, 0xb)
        /usr/src/go/src/pkg/net/fd_unix.go:419 +0x343
net.(*TCPListener).AcceptTCP(0xc20803a0b8, 0x4dae33, 0x0, 0x0)
        /usr/src/go/src/pkg/net/tcpsock_posix.go:234 +0x5d
net/http.tcpKeepAliveListener.Accept(0xc20803a0b8, 0x0, 0x0, 0x0, 0x0)
        /usr/src/go/src/pkg/net/http/server.go:1947 +0x4b
net/http.(*Server).Serve(0xc2080044e0, 0x7f1489ffffb0, 0xc20803a0b8, 0x0, 0x0)
        /usr/src/go/src/pkg/net/http/server.go:1698 +0x91
net/http.(*Server).ListenAndServe(0xc2080044e0, 0x0, 0x0)
        /usr/src/go/src/pkg/net/http/server.go:1688 +0x14d
net/http.ListenAndServe(0x776410, 0x5, 0x0, 0x0, 0x0, 0x0)
        /usr/src/go/src/pkg/net/http/server.go:1778 +0x79
main.serveHTTP()
        /go/src/app/main.go:36 +0x82
created by main.main
        /go/src/app/main.go:105 +0x2b8

goroutine 27 [IO wait]:
net.runtime_pollWait(0x7f1489fffbd0, 0x72, 0x0)
        /usr/src/go/src/pkg/runtime/netpoll.goc:146 +0x66
net.(*pollDesc).Wait(0xc20802bf00, 0x72, 0x0, 0x0)
        /usr/src/go/src/pkg/net/fd_poll_runtime.go:84 +0x46
net.(*pollDesc).WaitRead(0xc20802bf00, 0x0, 0x0)
        /usr/src/go/src/pkg/net/fd_poll_runtime.go:89 +0x42
net.(*netFD).readFrom(0xc20802bea0, 0x7f1489e5fd38, 0x200, 0x200, 0x0, 0x0, 0x0, 0x7f1489ffe520, 0xb)
        /usr/src/go/src/pkg/net/fd_unix.go:269 +0x3db
net.(*UDPConn).ReadFromUDP(0xc20803a0a8, 0x7f1489e5fd38, 0x200, 0x200, 0xc2080a2000, 0x0, 0x0, 0x0)
        /usr/src/go/src/pkg/net/udpsock_posix.go:67 +0x129
main.(*StatsDListener).Listen(0xc20803a0b0, 0xc2080a2000)
        /go/src/app/bridge.go:288 +0x8b
created by main.main
        /go/src/app/main.go:116 +0x422

goroutine 28 [select, 8 minutes]:
main.watchConfig(0x7fff43d6af43, 0xd, 0xc20800f200)
        /go/src/app/main.go:77 +0x766
created by main.main
        /go/src/app/main.go:124 +0x5bb

goroutine 30 [syscall, 8 minutes]:
syscall.Syscall(0x0, 0x5, 0xc208125e98, 0x10000, 0x0, 0x0, 0x0)
        /usr/src/go/src/pkg/syscall/asm_linux_amd64.s:21 +0x5
syscall.read(0x5, 0xc208125e98, 0x10000, 0x10000, 0x0, 0x0, 0x0)
        /usr/src/go/src/pkg/syscall/zsyscall_linux_amd64.go:838 +0x75
syscall.Read(0x5, 0xc208125e98, 0x10000, 0x10000, 0x0, 0x0, 0x0)
        /usr/src/go/src/pkg/syscall/syscall_unix.go:136 +0x5c
github.com/howeyc/fsnotify.(*Watcher).readEvents(0xc2080041e0)
        /go/src/github.com/howeyc/fsnotify/fsnotify_linux.go:219 +0x133
created by github.com/howeyc/fsnotify.NewWatcher
        /go/src/github.com/howeyc/fsnotify/fsnotify_linux.go:126 +0x288

goroutine 31 [chan receive, 8 minutes]:
github.com/howeyc/fsnotify.(*Watcher).purgeEvents(0xc2080041e0)
        /go/src/github.com/howeyc/fsnotify/fsnotify.go:21 +0x51
created by github.com/howeyc/fsnotify.NewWatcher
        /go/src/github.com/howeyc/fsnotify/fsnotify_linux.go:127 +0x2a0
```

**Comment by juliusv** [comment]
It crashes in this line:

https://github.com/prometheus/statsd_bridge/blob/master/bridge.go#L217

That probably means that NSQ sends counter increments with negative values (which Prometheus forbids). Can you confirm? Does NSQ internally "reset" its counters by decrementing them by their current value?

I agree the statsd-bridge shouldn't crash in that case, but probably should just ignore negative counter increments.

**Comment by ekristen** [comment]
I guess you could say they decrement by their current value, but really they just all reset to 0 upon a restart.

**Comment by juliusv** [comment]
The key thing is, they seem to send negative counter increments via the StatsD protocol. I'm not sure how much sense that makes. Usually you don't want a counter reset to be reflected as a negative rate (which I think will happen statsd-wise in this case). You'd just want to ignore it in the rate calculation. In any case, I'm all for simply ignoring negative counter increments in the bridge (while logging a warning about it). Would you be up for sending a PR?

**Comment by ekristen** [comment]
This is still a problem

**Comment by ekristen** [comment]
@juliusv if nsq restarts its stats reset, it isn't sending negative values, but it does send values that could be less than what it had originally.

**Comment by sdurrheimer** [comment]
Should this be closed given that #33 has been merged ?
